File: app/src/main/java/com/example/config/GlobalExceptionHandler.java
package com.example.config;

import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.HttpServerErrorException;
import java.nio.charset.StandardCharsets;
import java.net.URLEncoder;


@ControllerAdvice
@Slf4j
public class GlobalExceptionHandler {

    @ExceptionHandler(HttpServerErrorException.class)
    public String handleServerError(HttpServerErrorException e) {
        log.error("Server error from resource server: {}", e.getMessage(), e);
        String userMessage = "Произошла ошибка на сервере. Пожалуйста, попробуйте снова позже.";
        try {
            return "redirect:/error?message=" + URLEncoder.encode(userMessage, StandardCharsets.UTF_8);
        } catch (Exception ex) {
            log.error("Error encoding message: {}", ex.getMessage(), ex);
            return "redirect:/error?message=Unexpected+error";
        }
    }

    @ExceptionHandler(HttpClientErrorException.class)
    public String handleClientError(HttpClientErrorException e) {
        log.error("Client error from resource server: {}", e.getMessage(), e);
        String userMessage = switch (e.getStatusCode().value()) {
            case 400 -> "Некорректный запрос. Проверьте введённые данные.";
            case 401 -> "Требуется повторная авторизация.";
            case 403 -> "Доступ запрещён. У вас недостаточно прав.";
            default -> "Ошибка при выполнении запроса. Попробуйте снова.";
        };
        try {
            return "redirect:/error?message=" + URLEncoder.encode(userMessage, StandardCharsets.UTF_8);
        } catch (Exception ex) {
            log.error("Error encoding message: {}", ex.getMessage(), ex);
            return "redirect:/error?message=Unexpected+error";
        }
    }

}

File: app/src/main/kotlin/com/example/config/GlobalExceptionHandler.kt
package com.example.config

import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.HttpServerErrorException
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

@ControllerAdvice
class GlobalExceptionHandler {

    private val log = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(HttpServerErrorException::class)
    fun handleServerError(e: HttpServerErrorException): String {
        log.error("Server error from resource server: {}", e.message, e)
        val userMessage = "Resource server error. ${e.message}"
        return "redirect:/error?message=" + URLEncoder.encode(userMessage, StandardCharsets.UTF_8)
    }

    @ExceptionHandler(HttpClientErrorException::class)
    fun handleClientError(e: HttpClientErrorException): String {
        log.error("Client error from resource server: {}", e.message, e)
        val userMessage = when (e.statusCode.value()) {
            400 -> "Invalid request. Please check the data you entered."
            401 -> "Re-authentication required."
            403 -> "Access denied. You do not have sufficient permissions."
            else -> "An error occurred during the request. Please try again."
        }
        return "redirect:/error?message=" + URLEncoder.encode(userMessage, StandardCharsets.UTF_8)
    }
}
